pkg/ssl.go

```diff
@@ -0,0 +1,204 @@
+		"truststorePasswordConfirm": trustStorePassword,
+		"httpsHostname":             httpsHostname,
+		"httpsPort":                 httpsPort,
+	}
+
+	response, err := s.instance.http.
+		RequestFormData(params).
+		SetFiles(map[string]string{
+			"certificateFile": certificateFile,
+			"privatekeyFile":  privateKeyFile,
+		}).
+		Post(SSLSetupPath)
+
+	if err != nil {
+		return false, fmt.Errorf("%s > failed to setup SSL: %w", s.instance.ID(), err)
+	} else if response.IsError() {
+		// If the response is an error, we try to read the error message from the response body.
+		// The resty library does not provide a way to read the response body if the response is an error.
+		body, err := io.ReadAll(response.RawBody())
+		if err != nil {
+			return false, fmt.Errorf("%s > failed to setup SSL: %s, %w", s.instance.ID(), response.Status(), err)
+		}
+		errorMessage, err := findErrorMessage(string(body[:]))
+		if err != nil {
+			return false, fmt.Errorf("%s > failed to setup SSL: %s, %w", s.instance.ID(), response.Status(), err)
+		}
+		return false, fmt.Errorf("%s > failed to setup SSL: %s, %s", s.instance.ID(), response.Status(), errorMessage)
+	}
+
+	if err := lock.Lock(); err != nil {
+		return true, fmt.Errorf("%s > failed to lock SSL setup: %w", s.instance.ID(), err)
+	}
+
+	return true, nil
+}
+
+func writeDER(tempDerFile *os.File, pemBlock *pem.Block) error {
+	if _, err := tempDerFile.Write(pemBlock.Bytes); err != nil {
+		return err
+	}
+	err := tempDerFile.Close()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// From HTML response body, e.g.:
+// <!DOCTYPE html>
+// <html lang='en'>
+// <head>
+// <title>Error</title>
+// </head>
+// <body>
+// <h1>Error</h1>
+// <dl>
+// <dt class='foundation-form-response-status-code'>Status</dt>
+// <dd>500</dd>
+// <dt class='foundation-form-response-status-message'>Message</dt>
+// <dd>Invalid password for existing key store</dd>
+// <dt class='foundation-form-response-title'>Title</dt>
+// <dd>Error</dd>
+// </dl>
+// </body>
+// </html>
+// returns "Invalid password for existing key store"
+func findErrorMessage(body string) (errorMessage string, err error) {
+	doc, err := html.Parse(strings.NewReader(body))
+	if err != nil {
+		return "", err
+	}
+
+	var traverse func(*html.Node)
+	traverse = func(n *html.Node) {
+		if n.Type == html.ElementNode && n.Data == "dt" && hasClass(n, "foundation-form-response-status-message") {
+			// If the <dt> node with class 'foundation-form-response-status-message' is found,
+			// we search for the corresponding <dd> node containing the error message.
+			ddNode := n.NextSibling
+			for ddNode != nil {
+				if ddNode.Type == html.ElementNode && ddNode.Data == "dd" {
+					errorMessage = ddNode.FirstChild.Data
+					return
+				}
+				ddNode = ddNode.NextSibling
+			}
+		}
+		for c := n.FirstChild; c != nil; c = c.NextSibling {
+			traverse(c)
+		}
+	}
+
+	traverse(doc)
+	return errorMessage, nil
+}
+
+func hasClass(n *html.Node, class string) bool {
```

Comment:
to generic name for package `pkg`

maybe consider putting SSL utilities under `ssl` package?